De id chunking

shubham-s-agarwal opened the pull request:
Adding functionality for chunking documents that exceed the maximum number of tokens the model can process.
- Chunking done with straight cut, with measures taken to prevent sub-word division by implementing functionality to check for whitespace. 
- Added an attribute 'maximum_tokens_model' to the NER config to control the maximum number of tokens being passed (this has an upper limit of 510 to prevent issues down the road)

mart-r reviewed (changes requested):
This needs some improvement.

Please also include test cases such that it would fail without the chunking and it passes with chunking. Ideally, there would also be a test that fails by setting the token value higher in the config forcing everything to run in one batch (which is currently hindered by the hard-coded 510 limit).

This is what I came up with for a functionality test in `tests/utils/ner/test_deid.py`:
```python
# 2 copies should already be sufficient for text, really
# but doing 4 just in case
v_long_input_text = '\n\n'.join([input_text,]*4)


class DeIdModelWithLongText(unittest.TestCase):

    @classmethod
    def setUpClass(cls) -> None:
        _add_model(cls)
        cls.deid_model = train_model_once(cls.deid_model)[1]

    def test_works_for_long_text(self, text: str = v_long_input_text):
        anon_text = self.deid_model.deid_text(text)
        # NOTE: if the token limit is reached, it will identify
        #       and subsequently deidentify the first part of the
        #       text, but not the copies that follow.
        #       As such, the following will be in the text if
        #       the deid model doesn't do any batching.
        self.assertNotIn("Seen by Dr. M. Sully on 11/11/1996.", anon_text)
```

mart-r commented on medcat/utils/ner/deid.py:
```diff
@@ -62,18 +61,56 @@ def train(self, json_path: Union[str, list, None],
+
+        blocks = [[]]
+        whitespace_pattern = re.compile(r'\s')
+        maximum_tkns = min(510, config.general['maximum_tokens_model'])
```
If we're going to have a very specific logic in here, it would need to be documented.
And if it's known not to work with other types of models, we would need to detect and (at the very least) warn users if they are using something else.

tomolopolis replied:
agree - logger.warning & and comment here would be ideal if detected models from deserialized models detected to not be Roberta

tomolopolis reviewed (changes requested):
Logic is all there - but could use extraction of chunking / window logic to a new method on the class. Would the be easier to test in isolation. 
Some doc strings also of the tokenizer specific token_ids would also be good.

mart-r commented:
Would just like to point out (i.e I just noticed) that this change would currently not affect the multiprocessing method `deid_multi_texts`.

mart-r commented on medcat/utils/ner/deid.py:
```diff
@@ -62,18 +62,73 @@ def train(self, json_path: Union[str, list, None],
         return super().train(json_path, *args, train_nr=0, **kwargs)  # type: ignore
 
-    def deid_text(self, text: str, redact: bool = False) -> str:
+    def get_chunks(self, text:str, de_id_pipe,maximum_tkns) -> List[List[Any]]:
```
A few more things.
1. Please add type hints to arguments as well (and the linting issue).
2. The method should also have a docstring.
3. This method only depends on the tokenizer, not the entire pipe
4. I think this method should probably be extracted from the class, probably into its own module (i.e `chunking`) in the same package. After all, the method does not depend on the instance at all.
4. This method should simply return a list (or better yet, a generator) of strings (i.e the text chunks), not some obscure data structure.

mart-r commented on medcat/utils/ner/deid.py:
```diff
@@ -62,18 +62,73 @@ def train(self, json_path: Union[str, list, None],
+    def get_chunks(self, text:str, de_id_pipe,maximum_tkns) -> List[List[Any]]:
+
+        blocks = [[]]
+        if 'Roberta' in de_id_pipe.ner_pipe.tokenizer.__class__.__name__:
```
I think this is a good check, but I would extract the code below to a separate method, the name of which is explicit about it being for roberta models (i.e `get_chunks_roberta`) and a docstring that specifies that it's designed for these and is unlikely to work with anything else.

mart-r commented on medcat/utils/ner/deid.py:
```diff
@@ -62,18 +62,73 @@ def train(self, json_path: Union[str, list, None],
+                if len(blocks[-1]) == maximum_tkns:
+                    to_append_block = []
+                    idx_chunk = -1
+                    for i in range(len(blocks[-1]) - 1, len(blocks[-1]) - 21, -1):
```
I'm still not sure what this magic number of `21` is. From the comment below it seems like it's got something to do with sub-word division, but I don't know what.

mart-r commented on medcat/utils/ner/deid.py:
```diff
@@ -62,18 +62,73 @@ def train(self, json_path: Union[str, list, None],
+                    blocks[-1].append((token, (start, end)))
+
+        else:
+            # print("***********WARNING:************\nChunking functionality is implemented for RoBERTa model. The model detected is not RoBERTa, chunking omitted.\nPII information MAY BE REVEALED.")
```
Removed commented print line

mart-r commented on medcat/utils/ner/deid.py:
```diff
@@ -62,18 +62,73 @@ def train(self, json_path: Union[str, list, None],
+
+        else:
+            # print("***********WARNING:************\nChunking functionality is implemented for RoBERTa model. The model detected is not RoBERTa, chunking omitted.\nPII information MAY BE REVEALED.")
+            warnings.warn("\n\nChunking functionality is implemented for RoBERTa models. The detected model is not RoBERTa, so chunking is omitted. Be cautious, as PII information MAY BE REVEALED.")
```
Let us be consistent - create and use a module level `Logger` instead. Take a look at some of the other core modules to see how we generally do it.

mart-r commented on medcat/utils/ner/deid.py:
```diff
@@ -62,18 +62,73 @@ def train(self, json_path: Union[str, list, None],
+
+        if config is None:
+            config = ConfigTransformersNER()
+        maximum_tkns = min(510,config.general['maximum_tokens_model']) # This is specific to RoBERTa model, with a maximum token limit is 512.
```
If we want to specify an in code maximum of 510, I'd prefer if this was in a module level constant (and probably in the new `chunking` module). That way we can change it during testing (i.e to run a test that does not do chunking which thus fails to de-id).

mart-r reviewed (changes requested):
I still think we have a ways to go. Like Tom said, the logic is there, we just nee to make it all neat under the hood.

Plus, like I mentioned above:
- The deprecated `medcat.utils.ner.helpers.deid_text` method (used in tutorials and surely in some people's notebooks/code) will now result in a different result without chunking
  - We may remove it
  - Or we need to make sure it uses chunking (without duplicating code, of course)
- The `deid_multi_texts` method does not currently do any chunking

We still need to add tests.
IMHO the tests we need:
- Chunking tests
  - Check that joining chunks together returns original text
  - Check that chunking works for small documents (which require no chunking)
  - [Optional] Check that there's no chunking with non-roberta models
- Deid tests
  - Check that deid fails if the text is long and max tokens is long
  - Check that deid successful for long pieces of text that need chuking (with regular settings)

tomolopolis commented:
On the couple of issues:
- `deid_multi_texts`  - feel free to move this into another PR / add a log warning that this doesn't chunk - if you want to get to this later
- On the  `deid_text` deprecated module level method - doesn't necessarily need to implement chunking. Amend the warning of deprecation saying that this doesn't chunk

shubham-s-agarwal commented:
> On the couple of issues:
> 
> * `deid_multi_texts`  - feel free to move this into another PR / add a log warning that this doesn't chunk - if you want to get to this later
> * On the  `deid_text` deprecated module level method - doesn't necessarily need to implement chunking. Amend the warning of deprecation saying that this doesn't chunk

-  `deid_multi_texts` - I've added a log warning saying this doesn't perform chunking, PII data may be exposed
- `deid_text` deprecated module level method - I've added the chunking functionality which is similar to what we are doing for deid; I've added a parameter that signifies whether or not to chunk on the deprecated module level method.
